File: longest_substring.java
class Solution {
    // Runtime: O(n) - iterate through the whole list once
    // Memory: O(1)
    
    // Finds the length of the longest substring
    // with non-repeating characters
    public int lengthOfLongestSubstring(String s) {
        int i = 0;
        int j = 0;
        int maxLength = 0;
        HashSet<Character> chars = new HashSet();
        
        while (j < s.length()){
            char c = s.charAt(j);
            if(!chars.contains(c)){
                chars.add(c);
                // increment right pointer
                j++;

                // All set values are unique,
                maxLength = Math.max(chars.size(), maxLength);
            }else{
                // Remove the character left pointer is pointing to
                // Note: slides window one to the right
                chars.remove(s.charAt(i));
                // Increment left pointer
                i++;
            }
        }
        return maxLength;
    }
}

File: longestSubstring.py
# Problem: Longest SubString (Solved using sliding window technique)
# Runtime: O(n)
# Memory : O(n)
def lengthOfLongestSubString(s):
    i = 0  # Left corner of sliding window
    j = 0  # Right corner of sliding window
    seenChars = set()
    maxLength = 0
    for char in s:
        # If char isn't in set,
        # Increase the sliding window from the right
        if char not in seenChars:
            seenChars.add(s)
            j = j + 1
        else:
            # If character is in set, increase
            # the sliding window from the left (remove character currently at left pointer)
            seenChars.remove(char)
            i = i + 1
            maxLength = max(len(seenChars), maxLength)
    return maxLength


if __name__ == "__main__":
    print("Hello World")
